Set CATTLE_SERVER_URL when running Rancher with Docker

aruiz14 opened the pull request:
<!-- This template is for Devs to give QA details before moving the issue To-Test -->
### Summary

Fixes #14559

### Occurred changes and/or fixed issues

* By providing a valid value for the `server-url` Setting from the beginning (instead of during first-login page), startup will be faster, allowing Fleet to install the Fleet Agent in the local cluster almost immediately, instead of after several minutes (see parent issue).

### Technical notes summary
Fleet requires a working `apiServerURL` (part of the `config` payload in the `fleet-controller` ConfigMap), which is taken from the `server-url` `Setting`. 
In fact, it's very important to pick the right value for this field (normally filled during the bootstrap, where it defaults to the URL being used to access the UI), as some workloads (pods) will use this value to access Rancher. This means that needs to be an URL that is accessible from inside the Kubernetes networking.

### Areas or cases that should be tested
The parent issue did not include specific jobs that were failing. Please review if any tests were temporarily disabled which may need to be restored after this fix.

### Areas which could experience regressions
<!-- Create a detailed list of areas to be analyzed which may be affected by the changes, which would require a prior research to avoid regressions. -->

### Screenshot/Video
<!-- Attach screenshot or video of the changes and eventual comparison if you find it necessary -->

### Checklist
- [X] The PR is linked to an issue and the linked issue has a Milestone, or no issue is needed
- [X] The PR has a Milestone <!-- The milestone should automatically be assigned if the linked issue has one, but does not always happen (incorrectly linked, issue has no milestone, etc) -->
- [X] The PR template has been filled out
- [X] The PR has been self reviewed <!-- There are no TODOs, no incorrect files in the PR, all the required files are there, no commented out code, etc-->
- [X] The PR has a reviewer assigned
- [X] The PR has automated tests or clear instructions for manual tests and the linked issue has appropriate QA labels, or tests are not needed
- [X] ~The PR has reviewed with UX and tested in light and dark mode, or there are no UX changes~
- [X] ~The PR has been reviewed in terms of Accessibility~

richard-cox commented on scripts/e2e-docker-start:
```diff
@@ -25,6 +25,7 @@ docker run -d --restart=unless-stopped -p 80:80 -p 443:443 ${VOLUME_ARGS} \
   -e CATTLE_UI_OFFLINE_PREFERRED=true \
   -e CATTLE_BOOTSTRAP_PASSWORD=password \
   -e CATTLE_PASSWORD_MIN_LENGTH=3 \
+  -e CATTLE_SERVER_URL=https://host.docker.internal:443 \
```
Two questions

- Are docker installs that rely on localhost/127.0.0.1 broken (cannot deploy fleet bundles to upstream cluster)?
- The root issue talks about a 7-8 minute delay and deployments working. That's caused by the server-url setting being updated and new certs being generated OR it's still generating the original certs?

aruiz14 replied:
> - Are docker installs that rely on localhost/127.0.0.1 broken (cannot deploy fleet bundles to upstream cluster)?

When I tested this locally without specifying a server URL, the bootstrap page would select the internal Docker IP as default value for this parameter. However, it won't be "persisted" until the bootstrap is completed, and may miss some part of the bootstrapping, hence delaying the configuration completion.

> - The root issue talks about a 7-8 minute delay and deployments then working. That's caused by the server-url setting being updated and new certs being generated OR it's still generating the original certs?

TL;DR: By providing a working parameter since the beginning, startup is much faster, allowing Fleet to install the Fleet Agent in the local cluster almost immediately, instead of after several minutes.

Longer version: manager (Rancher) initiated agent registration requires a valid (reachable) server URL so Fleet can correctly configure the Fleet Agent to connect to the upstream cluster. There are 2 options:
1. If Rancher is running in Kubernetes, it will find the right host [here](https://github.com/rancher/rancher/blob/93a426631d58c7595c5d1e64dcff328f836cece5/pkg/controllers/provisioningv2/fleetcluster/fleetcluster.go#L214-L218).
2. As a fallback, it will pick the value from the `server-url` setting, which is what's used in Docker installations.

I believe the cert generation hasn't changed, but maybe the update of the setting is not immediately picked by Rancher/Fleet, hence introducing a delay. I _think_ I saw some exponential backoff retries in some of the affected code as well, so that could explain why the previous deployment time vary.

richard-cox replied:
default server-url

If the server-url isn't set when setup page loads then the UI defaults it the domain used to access the ui, so if used the docker ip it would insert that. The e2e tests grab it the same way. Mine always sets to `localhost:8081`.

cert delay causing bundle creation to be slow (or break?)

I speed ran a docker install --> setup --> bundle creation and didn't see a failure due to certs. Is the cert delay issue something specific to when the server-url is set to 127.0.0.1?

Something just doesn't feel right about this, like we're working around something instead of fixing the root issue

aruiz14 commented:
> This need to fix an issue in the dashboard repo, so please create one.
The parent issue was originally created in the dashboard repo, later moved to rancher/rancher. I've asked @gaktive to move it back to dashboard, since there is no point in having those, as there is no actual problem on the Fleet side, only a misconfiguration on the Rancher installation.

> Can you add some tests for this - our existing e2e tests have been passing without this change - is a change incoming from Fleet which will break that, which this fixes?
The original issue reported a problem with E2E continuously failing due to this. Since it was open for a few months, I suspect some of those tests may have been temporarily skipped to bypass those failures. That's why I asked in the description for someone with more knowledge about the test suite to check if that was the case an enable back such tests.
I think the value of this change resides on existing E2E continuing to work well. I'm also not sure which kind of additional tests could be added for this.

aruiz14 commented:
> I added a comment to the thread, but bringing up some of those points here and added new
> 
>     * found the affected e2e tests - cypress/e2e/tests/pages/fleet/resources/bundles.spec.ts

Thanks! I've just re-enabled those checks. My previous search didn't match due to the issue being moved between projects.

>     * i think the root issue is the delay in sorting out certs, however this might be just for 127.0.0.1 address? (see [here](https://github.com/rancher/dashboard/pull/14555#discussion_r2163277691))
> 
>     * this tweak feels like a workaround for a rancher issue, rather than a fleet or e2e config one. taking several minutes to create a cert doesn't feel right

In my tests, the certs were actually created. What takes a long time is setting the server URL. It's not "several minutes", it's actually "until the bootstrap is completed". In my tests, I ran Rancher and never logged in: the `server-url` was never set and Fleet wouldn't create the `fleet-agent` deployment in the `cattle-local-fleet-system` namespace.

>     * we have a very old [PR](https://github.com/rancher/dashboard/pull/10897) to use helm instead of docker in e2e tests, however requires some time to address failures (uses `kubectl cp` to patch ui files)

I wouldn't mind taking a look when I have some spare time if you think it's worth it.

richard-cox commented on scripts/e2e-docker-start:
```diff
@@ -21,12 +21,18 @@ fi
 
+# Configure the Rancher server URL from start, similarly to how chart installations work
+# The special host.docker.internal domain only works for Docker Desktop, not plain Docker Engine on linux, so using the special IP 172.17.0.1 to access the gateway of the default bridge network
+SERVER_URL_ARGS="-e CATTLE_SERVER_URL=https://172.17.0.1"
```
Can this just be `SERVER_URL="https://172.17.0.1"`, just to follow file convention

richard-cox commented on scripts/e2e-docker-start:
```diff
@@ -21,12 +21,18 @@ fi
 # "oidc-provider" feature flag needs a system restart so it's not feasible to enable it as part of the test
-docker run -d --restart=unless-stopped -p 80:80 -p 443:443 ${VOLUME_ARGS} \
+
+docker run -d --restart=unless-stopped -p 80:80 -p 443:443 -p 6060:6060 ${VOLUME_ARGS} \
```
can we pull out `-p 80:80 -p 443:443 -p 6060:6060` as `PORT_ARGS` and add a comment on why 6060 is needed?

richard-cox reviewed (changes requested):
Given the underlying issue was found and addressed...  this then becomes more about standardising the env and becomes a lot more palatable. with the two super minor tweaks i think we can get this in
